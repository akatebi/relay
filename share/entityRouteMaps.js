import {
  entityTypeToRevisionRoute,
  entityTypeToFamilyRoute,
  entityTypeToSimpleRoute,
  entityTypeToTreeNode,
  entityLabelToFamilyRoute,
  entityListLabels,
  entityTypeToLabel,
  entityTypeToPluralLabel,
  revisionableList,
  routeToEntityType,
} from './entityTypes';

// const debug = require('debug')('app:share:entityRouteMaps');

// this returns route that returns the ResultSetVM
export const entityRevRouteMap = (entityType) => {
  const route = entityTypeToRevisionRoute[entityType];
  // debug('entityRevRouteMap', entityType, route);
  if (!route) {
    const msg = `entityRevRouteMap(${entityType}) => undefined`;
    console.error(msg);
  }
  return route;
};

export const entityFamilyRouteMap = (entityType) => {
  const route = entityTypeToFamilyRoute[entityType];
  // debug('entityFamilyRouteMap', entityType, route);
  if (!route) {
    const msg = `entityFamilyRouteMap(${entityType}) => undefined`;
    console.error(msg);
  }
  return route;
};

// this returns route that returns an EntityVM list
export const entityRevSimpleMap = (entityType) => {
  const route = entityTypeToSimpleRoute[entityType];
  if (!route) {
    const msg = `entityRevSimpleMap(${entityType}) => undefined`;
    console.error(msg);
  }
  return route;
};

export const entityTreeNodeRouteMap = (entityType) => {
  // console.log('entityTreeNodeRouteMap', entityType);
  const route = entityTypeToTreeNode[entityType];
  if (!route) {
    const msg = `entityTreeNodeRouteMap(${entityType}) => undefined`;
    console.error(msg);
  }
  return route;
};

export const entityCreationFamilyRouteMap = (entityLabel) => {
  const route = entityLabelToFamilyRoute[entityLabel];
  if (!route) {
    const msg = `entityLabelFamily(${entityLabel}) => undefined`;
    console.error(msg);
  }
  return route;
};

export const entityRouteToTypeMap = (pathname) => {
  const entityRevRoute = pathname.split('/')[1];
  const entityType = routeToEntityType[entityRevRoute];
  if (!entityType) {
    const msg = `routeToEntityType(${entityRevRoute}) => undefined`;
    console.error(msg);
  }
  return entityType;
};

export const entityTypeToTreeNodeTypeMap = (entityType) => {
  const treeNodeType = entityTypeToTreeNode[entityType];
  if (!treeNodeType) {
    const msg = `routeToEntityType(${entityType}) => undefined`;
    console.error(msg);
  }
  return treeNodeType;
};

export const getEntityLabelByType = (entityType) => {
  const label = entityTypeToLabel[entityType];
  if (!label) {
    const msg = `entityListLabels has no "${entityType}"`;
    console.error(msg);
  }
  return label;
};

export const getEntityListLabel = () => {
  const entity = location.pathname.slice(1).split('/')[0];
  const key = Object.keys(entityListLabels).find(label =>
    new RegExp(label).test(entity));
  return key ? entityListLabels[key] : 'Unknown Entity';
};

export const getEntityPluralLabelByType = (entityType) => {
  const label = entityTypeToPluralLabel[entityType];
  if (!label) {
    const msg = `getEntityPluralLabelByType(${entityType}) => undefined`;
    console.error(msg);
  }
  return label;
};

export const getEntityPluralLabelByRoute = (pathname) => {
  const entityType = entityRouteToTypeMap(pathname);
  return getEntityPluralLabelByType(entityType);
};

export const isRevisionable = entity => revisionableList.includes(entity);

export const getEntityUrl = entityType =>
  (/LifeCycleConfig|User/.test(entityType) ?
    entityFamilyRouteMap(entityType) :
    entityRevRouteMap(entityType));
