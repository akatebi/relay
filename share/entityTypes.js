
export const EntityTypes = [
  'Action',
  'ApplicationRole',
  'ApprovalCycleConfig',
  'AssociationConfig',
  'CategoryHierarchy',
  'CharacteristicConfig',
  'ChoiceList',
  'DictionaryEntry',
  'Document',
  'DocumentClass',
  'DocumentType',
  'Layout',
  'LifeCycleConfig',
  'LineItemConfig',
  'NavigationHierarchy',
  'NavigationNode',
  'OrganizationHierarchy',
  'Organization',
  'SectionConfig',
  'User',
];

// EntityType ===>> Revision Route
export const entityTypeToRevisionRoute = {
  Action: 'actions/revisions',
  ApplicationRole: 'applicationroles/revisions',
  ApprovalCycleConfig: 'approvalcycleconfigs/revisions',
  AssociationConfig: 'associationconfigs/revisions',
  CategoryHierarchy: 'categoryhierarchies/revisions',
  CharacteristicConfig: 'characteristicconfigs/revisions',
  ChoiceList: 'choicelists/revisions',
  DictionaryEntry: 'dictionaryentries/revisions',
  Document: 'documents/revisions',
  DocumentClass: 'documentclasses/revisions',
  DocumentType: 'documenttypes/revisions',
  Layout: 'templatelayouts/revisions',
  LifeCycleAction: 'lifecycles/actions',
  LifeCycleConfig: 'lifecycleconfigs',
  LineItemConfig: 'lineitemconfigs/revisions',
  NavigationHierarchy: 'navigationhierarchies/revisions',
  NavigationNode: 'navigationnodes/revisions',
  OrganizationHierarchy: 'organizationhierarchies/revisions',
  Organization: 'organizations/revisions',
  SectionConfig: 'sectionconfigs/revisions',
  User: 'users',
};

// EntityType ===>> Family Route
export const entityTypeToFamilyRoute = {
  Action: 'actions',
  ApplicationRole: 'applicationroles',
  ApprovalCycleConfig: 'approvalcycleconfigs',
  AssociationConfig: 'associationconfigs',
  CategoryHierarchy: 'categoryhierarchies',
  CharacteristicConfig: 'characteristicconfigs',
  ChoiceList: 'choicelists',
  DictionaryEntry: 'dictionaryentries',
  Document: 'documents',
  DocumentClass: 'documentclasses',
  DocumentType: 'documenttypes',
  Layout: 'templatelayouts',
  LifeCycleAction: 'lifecycles/actions',
  LifeCycleConfig: 'lifecycleconfigs',
  LineItemConfig: 'lineitemconfigs',
  NavigationHierarchy: 'navigationhierarchies',
  NavigationNode: 'navigationnodes',
  OrganizationHierarchy: 'organizationhierarchies',
  Organization: 'organizations',
  SectionConfig: 'sectionconfigs',
  User: 'users',
};

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////

export const entityTypeToPluralLabel = {
  Action: 'Actions',
  ApplicationRole: 'Application Roles',
  ApprovalCycleConfig: 'Approval Cycle Configurations',
  AssociationConfig: 'Association Configurations',
  CharacteristicConfig: 'Characteristic Configurations',
  CategoryHierarchy: 'Category Hierarchies',
  ChoiceList: 'Choice Lists',
  Culture: 'Cultures',
  DictionaryEntry: 'Dictionary Entries',
  Document: 'Documents',
  DocumentClass: 'Document Classes',
  DocumentType: 'Document Types',
  Layout: 'Template Layouts',
  LifeCycleConfig: 'Life Cycle Configurations',
  LineItemConfig: 'Line Item Configurations',
  NavigationHierarchy: 'NavigationHierarchies',
  NavigationNode: 'NavigationNodes',
  OrganizationHierarchy: 'Organization Hierarchies',
  Organization: 'Organizations',
  SectionConfig: 'Section Configurations',
  SectionLineItemConfig: 'Section Line Item Configurations',
  User: 'User Profiles',
};

export const entityListLabels = {
  Action: 'Action',
  ApplicationRole: 'Application Role',
  ApprovalCycleConfig: 'Approval Cycle Configuration',
  Assignment: 'Assignment',
  AssociationConfig: 'Association Configuration',
  CategoryHierarchy: 'Category Hierarchy',
  CharacteristicConfig: 'Characteristic Configuration',
  ChoiceList: 'Choice List',
  DictionaryEntry: 'Dictionary Entry',
  Culture: 'Culture',
  Document: 'Document',
  DocumentClass: 'Document Class',
  DocumentType: 'Document Type',
  LifeCycleConfig: 'Life Cycle Configuration',
  LineItemConfig: 'Line Item Configuration',
  NavigationNode: 'Navigation Node',
  NavigationHierarchy: 'Navigation Hierarchy',
  Organization: 'Organization',
  OrganizationHierarchy: 'Organization Hierarchy',
  SectionConfig: 'Section Configuration',
  TemplateLayout: 'Template Layout',
  User: 'User Profile',
};

// EntityType ===>> Simple Route
export const entityTypeToSimpleRoute = {
  Action: 'actions/simple',
  ApplicationRole: 'applicationroles/simple',
  ApprovalCycleConfig: 'approvalcycleconfigs/simple',
  AssociationConfig: 'associationconfigs/simple',
  CategoryHierarchy: 'categoryhierarchies/simple',
  CharacteristicConfig: 'characteristicconfigs/simple',
  ChoiceList: 'choicelists/simple',
  DictionaryEntry: 'dictionaryentries/simple',
  Document: 'documents/simple',
  DocumentClass: 'documentclasses/simple',
  DocumentType: 'documenttypes/simple',
  Layout: 'templatelayouts/simple',
  LifeCycleAction: 'lifecycles/actions/simple',
  LifeCycleConfig: 'lifecycles/simple',
  LineItemConfig: 'lineitemconfigs/simple',
  NavigationHierarchy: 'navigationhierarchies/simple',
  NavigationNode: 'navigationnodes/simple',
  OrganizationHierarchy: 'organizationhierarchies/simple',
  Organization: 'organizations/simple',
  SectionConfig: 'sectionconfigs/simple',
  User: 'users/simple',
};

// EntityLabel ===>> Family Route
export const entityLabelToFamilyRoute = {
  Action: 'actions',
  'Application Role': 'applicationroles',
  'Approval Cycle Config': 'approvalcycleconfigs',
  'Association Config': 'associationconfigs',
  'Category Hierarchy': 'categoryhierarchies',
  'Characteristic Config': 'characteristicconfigs',
  'Choice List': 'choicelists',
  'Dictionary Entry': 'dictionaryentries',
  Document: 'documents',
  'Document Class': 'documentclasses',
  Layout: 'templatelayouts',
  'Lifecycle Config': 'lifecycleconfigs',
  'Line Item Config': 'lineitemconfigs',
  'Navigation Hierarchy': 'navigationhierarchies',
  'Navigation Node': 'navigationnodes',
  Organization: 'organizations',
  'Section Config': 'sectionconfigs',
  User: 'users',
};

// Route ===>> Entity Type
export const routeToEntityType = {
  actions: 'Action',
  applicationroles: 'ApplicationRole',
  approvalcycleconfigs: 'ApprovalCycleConfig',
  associationconfigs: 'AssociationConfig',
  categoryhierarchies: 'CategoryHierarchy',
  characteristicconfigs: 'CharacteristicConfig',
  choicelists: 'ChoiceList',
  dictionaryentries: 'DictionaryEntry',
  documents: 'Document',
  documentclasses: 'DocumentClass',
  documenttypes: 'DocumentType',
  lifecycleconfigs: 'LifeCycleConfig',
  lineitemconfigs: 'LineItemConfig',
  navigationhierarchies: 'NavigationHierarchy',
  navigationnodes: 'NavigationNode',
  organizationhierarchies: 'OrganizationHierarchy',
  organizations: 'Organization',
  sectionconfigs: 'SectionConfig',
  templatelayouts: 'TemplateLayout',
  users: 'User',
};

// Entity Type ===>> Tree Node Type
export const entityTypeToTreeNode = {
  CategoryHierarchy: 'dictionaryentries',
  ChoiceList: 'dictionaryentries',
  NavigationHierarchy: 'navigationnodes',
  OrganizationHierarchy: 'organizations',
};


export const entityTypeToLabel = {
  Action: 'Action',
  ApprovalCycleConfig: 'Approval Cycle Configuration',
  ApplicationRole: 'Application Role',
  Assignment: 'Assignment',
  AssociationConfig: 'Association Configuration',
  CategoryHierarchy: 'Category Hierarchy',
  CharacteristicConfig: 'Characteristic Configuration',
  ChoiceList: 'Choice List',
  Culture: 'Culture',
  DictionaryEntry: 'Dictionary Entry',
  Document: 'Document',
  DocumentClass: 'Document Class',
  DocumentType: 'Document Type',
  LifeCycleConfig: 'Life Cycle Configuration',
  LineItemConfig: 'Line Item Configuration',
  NavigationHierarchy: 'Navigation Hierarchy',
  NavigationNode: 'Navigation Node',
  OrganizationHierarchy: 'Organization Hierarchy',
  Organization: 'Organization',
  SectionConfig: 'Section Configuration',
  SectionLineItemConfig: 'Section Line Item Configuration',
  TemplateLayout: 'Template Layout',
  User: 'User Profile',
};

export const supportedCreationEntities = [
  'Action',
  'Application Role',
  'Approval Cycle Config',
  'Association Config',
  'Category Hierarchy',
  'Characteristic Config',
  'Choice List',
  'Dictionary Entry',
  'Document',
  'Document Class',
  'Layout',
  'Line Item Config',
  'Navigation Node',
  'Navigation Hierarchy',
  'Organization',
  'Section Config',
  'Sequence',
  'User',
];

export const propertyTypes = [
  'Boolean',
  'Category',
  // 'ControlNumber',
  'Date',
  'Decimal',
  'Entity',
  'EntityList',
  'Float',
  'Integer',
  'RoleList',
  'RichText',
  'Text',
];

export const revisionableList = [
  'actions',
  'applicationroles',
  'approvalcycleconfigs',
  'associationconfigs',
  'categoryhierarchies',
  'characteristicconfigs',
  'choicelists',
  'dictionaryentries',
  'documents',
  'documentclasses',
  'documenttypes',
  'lineitemconfigs',
  'navigationhierarchies',
  'navigationnodes',
  'organizations',
  'organizationhierarchies',
  'sectionconfigs',
  'templatelayouts',
];
