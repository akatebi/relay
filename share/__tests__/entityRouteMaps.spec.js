import { getEntityUrl } from '../entityRouteMaps';
import { EntityTypes } from '../entityTypes';


describe('Enity Route Maps', () => {

  EntityTypes.forEach(entityType =>
    test(entityType, () => {
      expect(getEntityUrl(entityType)).toBeDefined();
      expect(getEntityUrl(entityType)).toMatchSnapshot();
    }));

});
