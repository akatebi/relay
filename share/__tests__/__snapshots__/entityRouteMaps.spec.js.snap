// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Enity Route Maps Action 1`] = `"actions/revisions"`;

exports[`Enity Route Maps ApplicationRole 1`] = `"applicationroles/revisions"`;

exports[`Enity Route Maps ApprovalCycleConfig 1`] = `"approvalcycleconfigs/revisions"`;

exports[`Enity Route Maps AssociationConfig 1`] = `"associationconfigs/revisions"`;

exports[`Enity Route Maps CategoryHierarchy 1`] = `"categoryhierarchies/revisions"`;

exports[`Enity Route Maps CharacteristicConfig 1`] = `"characteristicconfigs/revisions"`;

exports[`Enity Route Maps ChoiceList 1`] = `"choicelists/revisions"`;

exports[`Enity Route Maps DictionaryEntry 1`] = `"dictionaryentries/revisions"`;

exports[`Enity Route Maps Document 1`] = `"documents/revisions"`;

exports[`Enity Route Maps DocumentClass 1`] = `"documentclasses/revisions"`;

exports[`Enity Route Maps DocumentType 1`] = `"documenttypes/revisions"`;

exports[`Enity Route Maps Layout 1`] = `"templatelayouts/revisions"`;

exports[`Enity Route Maps LifeCycleConfig 1`] = `"lifecycleconfigs"`;

exports[`Enity Route Maps LineItemConfig 1`] = `"lineitemconfigs/revisions"`;

exports[`Enity Route Maps NavigationHierarchy 1`] = `"navigationhierarchies/revisions"`;

exports[`Enity Route Maps NavigationNode 1`] = `"navigationnodes/revisions"`;

exports[`Enity Route Maps Organization 1`] = `"organizations/revisions"`;

exports[`Enity Route Maps OrganizationHierarchy 1`] = `"organizationhierarchies/revisions"`;

exports[`Enity Route Maps SectionConfig 1`] = `"sectionconfigs/revisions"`;

exports[`Enity Route Maps User 1`] = `"users"`;
